src/main/java/ru/itmo/sd/deadliner2bot/messages/MessageFormatter.java

```diff
@@ -0,0 +1,47 @@
+package ru.itmo.sd.deadliner2bot.messages;
+
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.support.PropertiesLoaderUtils;
+import org.springframework.stereotype.Component;
+import ru.itmo.sd.deadliner2bot.dto.TodoDto;
+
+import javax.annotation.PostConstruct;
+import java.io.IOException;
+import java.util.*;
+
+@Component
+public class MessageFormatter {
+
+    private Properties messageTemplatesProperties;
+
+    @PostConstruct
+    public void postConstruct() throws IOException {
+        Resource resource = new ClassPathResource("message-templates/message-templates.properties");
+        messageTemplatesProperties = PropertiesLoaderUtils.loadProperties(resource);
+    }
+
+    public String dailyNotificationMessage(List<TodoDto> todoDtos) {
+        StringJoiner lines = new StringJoiner("\n");
+        lines.add(messageTemplatesProperties.getProperty("daily-notifications-header"));
+        todoDtos.stream()
+                .sorted(Comparator.nullsLast(Comparator.comparing(TodoDto::getEndTime)))
```

Comment:
If I am right, here list has been already sorted (we got sorted TreeSet from database, after this perform operations in TreeSet order). Anyway, let's keep invariant, that notifications are sorted everywhere include database.

Reply:
This class operates with DTOs and thus all ordering guarantees are missing. Better to sort before rendering, ordering doesn't matter before.

Reply:
From this point, I think we should replace Set in JPA classes by List

Reply:
You should investigate a bit more, using List is actually more expensive than set.

Reply:
Ok